src/Workspaces/CoreTest/SolutionTests/TryApplyChangesTests.cs

```diff
@@ -0,0 +1,169 @@
+
+            var exception = Assert.Throws<NotSupportedException>(
+                () => workspace.TryApplyChanges(project.WithCompilationOptions(project.CompilationOptions!.WithMainTypeName("WrongThing")).Solution));
+
+            Assert.Equal(WorkspacesResources.Changing_compilation_options_is_not_supported, exception.Message);
+        }
+
+        [Fact]
+        public void TryApplyWorksIfAllowingAnyParseOption()
+        {
+            // If we simply support the main change kind, then any type of change should be supported and we should not get called
+            // to the other method
+            using var workspace = new CustomizedCanApplyWorkspace(
+                allowedKinds: new[] { ApplyChangesKind.ChangeParseOptions },
+                canApplyParseOptions: (_, __) => throw new Exception("This should not have been called."));
+
+            var project = workspace.CurrentSolution.Projects.Single();
+
+            Assert.True(workspace.TryApplyChanges(
+                project.WithParseOptions(
+                    project.ParseOptions!.WithFeatures(new[] { KeyValuePairUtil.Create("Feature", "") })).Solution));
+        }
+
+        [Fact]
+        public void TryApplyWorksSpecificChangeIsAllowedForParseOption()
+        {
+            // If we don't support the main change kind, then the other method should be called
+            using var workspace = new CustomizedCanApplyWorkspace(
+                allowedKinds: new ApplyChangesKind[] { },
+                canApplyParseOptions: (_, newParseOptions) => newParseOptions.Features["Feature"] == "ExpectedValue");
+
+            var project = workspace.CurrentSolution.Projects.Single();
+
+            Assert.True(
+                workspace.TryApplyChanges(
+                    project.WithParseOptions(project.ParseOptions!.WithFeatures(new[] { KeyValuePairUtil.Create("Feature", "ExpectedValue") })).Solution));
+        }
+
+        [Fact]
+        public void TryApplyWorksThrowsIfChangeIsDisallowedForParseOption()
+        {
+            // If we don't support the main change kind, then the other method should be called
+            using var workspace = new CustomizedCanApplyWorkspace(
+                allowedKinds: new ApplyChangesKind[] { },
+                canApplyParseOptions: (_, newParseOptions) => newParseOptions.Features["Feature"] == "ExpectedValue");
+
+            var project = workspace.CurrentSolution.Projects.Single();
+
+            var exception = Assert.Throws<NotSupportedException>(
+                () => workspace.TryApplyChanges(
+                    project.WithParseOptions(project.ParseOptions!.WithFeatures(new[] { KeyValuePairUtil.Create("Feature", "WrongThing") })).Solution));
+
+            Assert.Equal(WorkspacesResources.Changing_parse_options_is_not_supported, exception.Message);
+        }
+
+        [Fact]
+        public void TryApplyWorksWhenAddingEditorConfigWithoutSupportingCompilationOptionsChanging()
```

Comment:
This may seem odd here, but this was to make sure that #44331 doesn't break this. This test originally started out in that PR, but it was clear that it's worth generally documenting and this standalone, and to also be able to have a direct conversation about how the behavior is here.